man/dpkg-buildpackage.1
.TH dpkg\-buildpackage 1 "2009-06-13" "Debian Project" "dpkg utilities"
.SH NAME
dpkg\-buildpackage \- build binary or source packages from sources
.
.SH SYNOPSIS
.B dpkg\-buildpackage
.RI [ options ]
.
.SH DESCRIPTION
.B dpkg\-buildpackage
is a program that automates the process of building a Debian package. It
consists of the following steps:
.IP \fB1.\fP 3
It prepares the build environment by setting various environment
variables (see \fBENVIRONMENT VARIABLES\fP).
.IP \fB2.\fP 3
It checks that the build-dependencies and build-conflicts
are satisfied (unless \fB\-d\fP is specified).
.IP \fB3.\fP 3
If a specific target has been selected with the \fB-T\fP or \fB--target\fP
option, it calls that target and stops here. Otherwise it calls
\fBfakeroot debian/rules clean\fP to clean the build-tree (unless
\fB\-nc\fP is specified).
.IP \fB4.\fP 3
It calls \fBdpkg-source\fP to generate the source package (unless
a binary-only build has been requested with \fB\-b\fP, \fB\-B\fP or
\fB\-A\fP).
.IP \fB5.\fP 3
It calls \fBdebian/rules\fP \fBbuild\fP followed by
\fBfakeroot debian/rules\fP \fIbinary-target\fP (unless a source-only
build has been requested with \fB\-S\fP). Note that \fIbinary-target\fR is
either \fBbinary\fP (default case, or if \fB\-b\fP is specified)
or \fBbinary-arch\fP (if \fB\-B\fP is specified) or \fBbinary-indep\fP
(if \fB\-A\fP is specified).
.IP \fB6.\fP 3
It calls \fBgpg\fP to sign the \fB.dsc\fP file (if any, unless
\fB\-us\fP is specified).
.IP \fB7.\fP 3
It calls \fBdpkg-genchanges\fP to generate a \fB.changes\fP file.
Many \fBdpkg-buildpackage\fP options are forwarded to
\fBdpkg-genchanges\fP.
.IP \fB8.\fP 3
It calls \fBgpg\fP to sign the \fB.changes\fP file (unless \fB\-uc\fP
is specified).
.IP \fB9.\fP 3
If \fB\-tc\fP is specified, it will call \fBfakeroot debian/rules clean\fP
again.
.
.SH OPTIONS
.TP
.B \-b
Specifies a binary-only build, no source files are to be built and/or
distributed. Passed to \fBdpkg-genchanges\fP.
.TP
.B \-B
Specifies a binary-only build, limited to architecture dependent packages.
Passed to \fBdpkg-genchanges\fP.
.TP
.B \-A
Specifies a binary-only build, limited to architecture independent
packages. Passed to \fBdpkg-genchanges\fP.
.TP
.B \-S
Specifies a source-only build, no binary packages need to be made.
Passed to \fBdpkg-genchanges\fP.
.P
.BI \-\-target= target
.br
.BI "\-\-target " target
.br
.BI \-T target
.RS
Calls \fBdebian/rules\fP \fItarget\fP after having setup the build
environment and stops the package build process here. If
\fB\-\-as\-root\fP is also given, then the command is executed
as root (see \fB\-r\fP). Note that official targets that are required to
be run as root by the Debian policy do not need this option.
.RE
.TP
.B \-\-as\-root
Only meaningful together with \fB\-\-target\fP. Requires that the target be
run with root rights.
.P
.B \-si
.br
.B \-sa
.br
.B \-sd
.br
.BI \-v version
.br
.BI \-C changesdescription
.br
.BI \-m maintaineraddress
.br
.BI \-e maintaineraddress
.RS
Passed unchanged to \fBdpkg-genchanges\fP. See its manual page.
.RE
.TP
.BI \-a architecture
Specify the Debian architecture we build for. The architecture of the
machine we build on is determined automatically, and is also the default
for the host machine.

If the host architecture differs from the build architecture (as is the
case for a cross-compilation), and if the environment variable
\fBPKG_CONFIG_LIBDIR\fP is not set, then it is set to a value suitable for
cross-compilation
("/usr/\fIgnu-system-type\fP/lib/pkgconfig/:/usr/share/pkgconfig").
.TP
.BI \-t gnu-system-type
Specify the GNU system type we build for. It can be used in place
of \-a or as a complement to override the default GNU system type
of the target Debian architecture.
.TP
.BI \-j jobs
Number of jobs allowed to be run simultaneously, equivalent to the
.BR make (1)
option of the same name. Will add itself to the MAKEFLAGS
environment variable, which should cause all subsequent make
invocations to inherit the option. Also adds \fBparallel=\fP\fIjobs\fP
to the DEB_BUILD_OPTIONS environment variable which allows
debian/rules files to use this information for their own purposes.
The \fBparallel=\fP\fIjobs\fP in DEB_BUILD_OPTIONS environment variable
will override the \fB-j\fP value if this option is given.
.TP
.B \-D
Check build dependencies and conflicts; abort if unsatisfied. This is the
default behavior.
.TP
.B \-d
Do not check build dependencies and conflicts.
.TP
.B \-nc
Do not clean the source tree (implies \fB\-b\fP if nothing else has been
selected among \fB-B\fP, \fB-A\fP or \fB-S\fP).
.TP
.B \-tc
Clean the source tree (using
.I gain-root-command
.BR "debian/rules clean" )
after the package has been built.
.TP
.BI \-r gain-root-command
When
.B dpkg\-buildpackage
needs to execute part of the build process as root, it prefixes the
command it executes with
.I gain-root-command
if one has been specified. Otherwise, if none has been specified,
\fBfakeroot\fP will be used by default, if the command is present.
.I gain-root-command
should start with the name of a program on the
.B PATH
and will get as arguments the name of the real command to run and the
arguments it should take.
.I gain-root-command
can include parameters (they must be space-separated) but no shell
metacharacters.
.I gain-root-command
might typically be
.BR fakeroot ", " sudo ", " super " or " really .
.B su
is not suitable, since it can only invoke the user's shell with
.B \-c
instead of passing arguments individually to the command to be run.
.TP
.BI \-R rules-file
Building a Debian package usually involves invoking
.B debian/rules
as a command with several standard parameters. With this option it's
possible to use another program invocation to build the package (it can
include space separated parameters).
Alternatively it can be used to execute the standard rules file with
another make program (for example by using
.B /usr/local/bin/make -f debian/rules
as \fIrules-file\fR).
.TP
.BI \-p sign-command
When \fBdpkg\-buildpackage\fP needs to execute GPG or PGP to sign a source
control (\fB.dsc\fP) file or a \fB.changes\fP file it will run
\fIsign-command\fP (searching the \fBPATH\fP if necessary) instead of
\fBgpg\fP. \fIsign-command\fP will get all the arguments that
\fBpgp\fP would have gotten. If \fIsign-command\fP
takes its arguments in GPG rather than PGP style, you should give
the \fB\-sgpg\fP option. \fIsign-command\fP
should not contain spaces or any other shell metacharacters.
.TP
.BI \-k key-id
Specify a key-ID to use when signing packages.
.TP
.BR \-us
Do not sign the source package.
.TP
.BR \-uc
Do not sign the \fB.changes\fP file.
.P
.BR \-i [\fIregexp\fP]
.br
.BI \-I [\fIpattern\fP]
.br
.BR \-s [ nsAkurKUR ]
.br
.BR \-z ", " \-Z
.br
.RS
Passed unchanged to \fBdpkg\-source\fP. See its manual page.
.RE
.P
.BI \-\-admindir= dir
.br
.BI "\-\-admindir " dir
.RS
Change the location of the \fBdpkg\fR database. The default location is
\fI/var/lib/dpkg\fP.
.RE
.TP
.BR \-h ", " \-\-help
Show the usage message and exit.
.TP
.BR \-\-version
Show the version and exit.
.
.SH ENVIRONMENT VARIABLES
.SS Variables set by dpkg-architecture
\fBdpkg\-architecture\fP is called with the \fB\-a\fP and \fB\-t\fP
parameters forwarded. Any variable that is output by its \fB\-s\fP
option is integrated in the build environment.
.SS Compiler flags
Some environment variables defining compiler and linker options are
set to default values unless already present in the environment. Note that
this mechanism was only introduced in version 1.14.17 of dpkg-dev and
not all \fIrules\fP files and build tools will honour these variables,
yet.
See https://wiki.ubuntu.com/DistCompilerFlags for the details.
.TP
.B CFLAGS
Optimization options which are passed to the Debian build system and
can/should be overriden by the package build if needed (default value:
.BR "\-g \-O2" ,
or
.BR \-g\ \-O0
if
.BR noopt
is specified in DEB_BUILD_OPTIONS). Overriding options can be
used to explicitely set a
higher optimization level, or work around compiler bugs, which only
can be seen with some optimization levels (the last opt level "wins").
.TP
.B CFLAGS_APPEND
Optimization options appended to the compiler flags, which must not be
overwritten by the package (mostly used to for test builds). Default
value: empty.
.TP
.B CXXFLAGS
Same as
.B CFLAGS
for C++ sources.
.TP
.B CXXFLAGS_APPEND
Same as
.B CFLAGS_APPEND
for C++ sources.
.TP
.B FFLAGS
Same as
.B CFLAGS
for Fortran sources.
.TP
.B FFLAGS_APPEND
Same as
.B CFLAGS_APPEND
for Fortran sources.
.TP
.B CPPFLAGS
Preprocessor flags which are passed to the Debian build system and
can/should be overriden by the package build if needed (default:
empty). This macro is seldom used (most build systems just use
.B CFLAGS
instead of
.BR CPPFLAGS ).
.TP
.B CPPFLAGS_APPEND
Preprocessor flags appended to the preprocessor flags, which must not
be overwritten by the package (mostly used to for test
builds). Default value: empty.
.TP
.B LDFLAGS
Options passed to the compiler when linking executables or shared
objects (if the linker is called directly, then
.B -Wl
and
.B ,
have to be stripped from these options). Default value: empty.
.TP
.B LDFLAGS_APPEND
Optimization options appended to the compiler flags when linking code,
which must not be overwritten by the package (mostly used to for test
builds). Default value: empty.
.TP
.B DEB_BUILD_OPTIONS
As indicated above,
.BR noopt
will disable optimation.
Additionally, when used with the
.BR hardening-wrapper
package, the values
.BR hardening
and
.BR nohardening
will be converted into their respective DEB_BUILD_HARDENING values.
The
.BR hardening
option can also include (optionally prefixed with
.BR no
) the following sub-options:
.BR stackprotector
.BR format
.BR fortify
.BR pie
.BR relro
For example,
.BR DEB_BUILD_OPTIONS=hardening=nopie
would cause
.BR DEB_BUILD_HARDENING_PIE=0
to be set, or
.BR DEB_BUILD_OPTIONS=nohardening
would cause

debian/changelog
dpkg (1.15.5.6ubuntu4.5+test1) lucid-proposed; urgency=low

  * Disable by default usage of synchronous sync(2), as it causes undesired
    I/O on unrelated file systems. Closes: #588339, #595927, #600075
  * On Linux use sync_file_range() to initiate asynchronous writeback
    of just unpacked files. Suggested by Ted Ts'o <tytso@mit.edu>.
    Thanks to Jonathan Nieder <jrnieder@gmail.com>. Closes: #605009

 -- Michael Jeanson <mjeanson@rlnx.com>  Mon, 06 Aug 2012 18:49:15 +0000

dpkg (1.15.5.6ubuntu4.5) lucid-security; urgency=low

  * SECURITY UPDATE: relative directory and symlink following in source pkgs.
    - scripts/Dpkg/Source/Archive.pm, scripts/Dpkg/Source/Patch.pm,
      scripts/Dpkg/Source/Package/V2.pm: applied fixes from Raphael Hertzog,
      thanks to Raphael Geissert.
    - CVE-2010-1679

 -- Kees Cook <kees@ubuntu.com>  Thu, 06 Jan 2011 11:08:21 -0800

dpkg (1.15.5.6ubuntu4.4) lucid-proposed; urgency=low

  * scripts/Dpkg/Source/Archive.pm: Drop -k parameter from the tar call used 
    by dpkg-source to extract tarballs. Upstream binary files modified by the 
    packager were not properly installed due to this.  Patch based on that of 
    Raphaël Hertzog (upstream, debian git). (LP: #633015)

 -- Dave Walker (Daviey) <DaveWalker@ubuntu.com>  Sat, 02 Oct 2010 00:30:53 +0100

dpkg (1.15.5.6ubuntu4.3) lucid-security; urgency=low

  * SECURITY UPDATE: no change rebuild to use the new statically linked libbz2
    which fixed CVE-2010-0405

 -- Jamie Strandboge <jamie@ubuntu.com>  Mon, 20 Sep 2010 07:56:11 -0500

dpkg (1.15.5.6ubuntu4.1) lucid-proposed; urgency=low

  * Backport proposed patch from Guillem Jover:
    - On Linux, call sync() (which is synchronous) before rename() rather
      than calling fsync() once per file (LP: #570805).

 -- Colin Watson <cjwatson@ubuntu.com>  Mon, 28 Jun 2010 14:32:02 +0100

dpkg (1.15.5.6ubuntu4) lucid; urgency=low

  * Backport from upstream:
    - Restore fsync during package unpack (LP: #559915).  This is now done
      by deferring the fsync and rename for normal files in tar extraction
      so that it's done in one pass afterwards, to avoid massive I/O
      degradation due to the serialization from each write + fsync.  When
      creating hard links to normal files on extraction use the .dpkg-new
      filename for source as the file is not yet in place due to the rename
      deferral.
    - Fix dpkg to not lose package metadata on filesystems where readdir()
      returns new files added after the opendir() call, btrfs in particular
      triggered the problematic behaviour. Closes: #575891
    - Report deferred trigger errors on status-fd. Closes: #574599,
      LP: #540252
      Thanks to Michael Vogt <michael.vogt@ubuntu.com>.

 -- Colin Watson <cjwatson@ubuntu.com>  Thu, 15 Apr 2010 12:38:50 +0100

dpkg (1.15.5.6ubuntu3) lucid; urgency=low

  * Revert fsync during package unpack for now; it's unacceptably slow for
    packages with lots of small files, and we can't ship beta-1 this way.
    We'll do something better once it's decided upstream (LP: #537241).

 -- Colin Watson <cjwatson@ubuntu.com>  Tue, 16 Mar 2010 10:04:38 +0000

dpkg (1.15.5.6ubuntu2) lucid; urgency=high

  * Backport from upstream:
    - Use FIEMAP when available (on Linux based systems) to sort the .list
      files loading order. With a cold cache it improves up to a 70%.
      Thanks to Morten Hustveit <morten@debian.org>. LP: #442114
    - Call fsync(2) after writing files on disk, to get the atomicity
      guarantees when doing rename(2). Based on a patch by Jean-Baptiste
      Lallement <jeanbaptiste.lallement@gmail.com>.
      Closes: #430958, LP: #512096
  * Security fixes by Raphaël Hertzog, also backported from upstream
    (CVE-2010-0396):
    - Modify dpkg-source to error out when it would apply patches containing
      insecure paths (with "/../") and also error out when it would apply a
      patch through a symlink. Those checks are required as patch will
      happily modify files outside of the target directory and unpacking a
      source package should not be able to have any side-effect outside of
      the target directory. LP: #532445
    - Also error out when the quilt series contains a path with "/../" as
      this can cause patch to create files outside of the source package due
      to the -B .pc/$path option that it gets.

 -- Colin Watson <cjwatson@ubuntu.com>  Thu, 11 Mar 2010 00:34:28 +0000

dpkg (1.15.5.6ubuntu1) lucid; urgency=low

  * Resynchronise with Debian.  Remaining changes:
    Ubuntu-specific adjustments (probably):
    - Adjust versioned emacs22 conflicts to cope with versions in Ubuntu.

    Miscellaneous bug fixes:
    - Avoid duplicate attempts to [f]close in obscure error situations which
      might conceivably close wrong fds.
    - Revert change to stop outputting a newline after a postinst is run
      (Debian #392317).

    Build options:

scripts/dpkg-buildpackage.pl
#!/usr/bin/perl
#
# dpkg-buildpackage
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

use strict;
use warnings;

use Cwd;
use File::Basename;
use POSIX;

use Dpkg;
use Dpkg::Gettext;
use Dpkg::ErrorHandling;
use Dpkg::BuildOptions;
use Dpkg::Compression;
use Dpkg::Version;
use Dpkg::Changelog::Parse;
use Dpkg::Arch qw(get_build_arch debarch_to_gnutriplet);

textdomain("dpkg-dev");

sub showversion {
    printf _g("Debian %s version %s.\n"), $progname, $version;

    print _g("
Copyright (C) 1996 Ian Jackson.
Copyright (C) 2000 Wichert Akkerman
Copyright (C) 2007 Frank Lichtenheld");

    print _g("
This is free software; see the GNU General Public Licence version 2 or
later for copying conditions. There is NO warranty.
");
}

sub usage {
	printf _g("
Usage: %s [<options> ...]

Options:
  -r<gain-root-command>
                 command to gain root privileges (default is fakeroot).
  -R<rules>      rules file to execute (default is debian/rules).
  -p<sign-command>
  -d             do not check build dependencies and conflicts.
  -D             check build dependencies and conflicts.
  -T<target>     call debian/rules <target> with the proper environment
  --as-root      ensure -T calls the target with root rights
  -j[<number>]   specify jobs to run simultaneously } passed to debian/rules
  -k<keyid>      the key to use for signing.
  -sgpg          the sign-command is called like GPG.
  -spgp          the sign-command is called like PGP.
  -us            unsigned source.
  -uc            unsigned changes.
  -a<arch>       Debian architecture we build for (implies -d).
  -b             binary-only, do not build source.   } also passed to
  -B             binary-only, no arch-indep files.   } dpkg-genchanges
  -A             binary-only, only arch-indep files. }
  -S             source only, no binary files.     }
  -t<system>     set GNU system type.           } passed to dpkg-architecture
  -v<version>    changes since version <version>.      }
  -m<maint>      maintainer for package is <maint>.    }
  -e<maint>      maintainer for release is <maint>.    } only passed
  -C<descfile>   changes are described in <descfile>.  } to dpkg-genchanges
  -si (default)  src includes orig if new upstream.    }
  -sa            uploaded src always includes orig.    }
  -sd            uploaded src is diff and .dsc only.   }
  -sn            force Debian native source format.      }
  -s[sAkurKUR]   see dpkg-source for explanation.        } only passed
  -z<level>      compression level of source             } to dpkg-source
  -Z<compressor> compression to use for source           }
  -nc            do not clean source tree (implies -b).
  -tc            clean source tree when finished.
  -ap            add pause before starting signature process.
  -i[<regex>]    ignore diffs of files matching regex.    } only passed
  -I[<pattern>]  filter out files when building tarballs. } to dpkg-source
  --admindir=<directory>
                 change the administrative directory.
  -h, --help     show this help message.
      --version  show the version.
"), $progname;
}

my @debian_rules = ("debian/rules");
my @rootcommand = ();
my $signcommand = '';
if ( ( ($ENV{GNUPGHOME} && -e $ENV{GNUPGHOME})
       || ($ENV{HOME} && -e "$ENV{HOME}/.gnupg") )
     && testcommand('gpg')) {
	 $signcommand = 'gpg';
} elsif (testcommand('pgp')) {
	$signcommand = 'pgp'
}

my ($admindir, $signkey, $forcesigninterface, $usepause, $noclean,
    $sourcestyle, $cleansource,
    $binaryonly, $sourceonly, $since, $maint,
    $changedby, $desc, $parallel);
my (@checkbuilddep_args, @passopts, @tarignore);
my $checkbuilddep = 1;
my $signsource = 1;
my $signchanges = 1;
my $diffignore = '';
my $binarytarget = 'binary';
my $targetarch = my $targetgnusystem = '';
my $call_target = '';
my $call_target_as_root = 0;

while (@ARGV) {
    $_ = shift @ARGV;

    if (/^(--help|-h)$/) {
	usage;
	exit 0;
    } elsif (/^--version$/) {
	showversion;
	exit 0;
    } elsif (/^--admindir$/) {
        $admindir = shift @ARGV;
    } elsif (/^--admindir=(.*)$/) {
	$admindir = $1;
    } elsif (/^-j(\d*)$/) {
	$parallel = $1 || '';
    } elsif (/^-r(.*)$/) {
	@rootcommand = split /\s+/, $1;
    } elsif (/^-p(.*)$/) {
	$signcommand = $1;
    } elsif (/^-k(.*)$/) {
	$signkey = $1;
    } elsif (/^-([dD])$/) {
	$checkbuilddep = ($1 eq 'D');
    } elsif (/^-s(gpg|pgp)$/) {
	$forcesigninterface = $1;
    } elsif (/^-us$/) {
	$signsource = 0;
    } elsif (/^-uc$/) {
	$signchanges = 0;
    } elsif (/^-ap$/) {
	$usepause = 1;
    } elsif (/^-a(.*)$/) {
	$targetarch = $1;
	$checkbuilddep = 0;
    } elsif (/^-s[iad]$/) {
	$sourcestyle = $_;
    } elsif (/^-s[nsAkurKUR]$/) {
	push @passopts, $_; # passed to dpkg-source
    } elsif (/^-[zZ]/) {
	push @passopts, $_; # passed to dpkg-source
    } elsif (/^-i.*$/) {
	$diffignore = $_;
    } elsif (/^-I.*$/) {
	push @tarignore, $_;
    } elsif (/^-tc$/) {
	$cleansource = 1;
    } elsif (/^-t(.*)$/) {
	$targetgnusystem = $1; # Order DOES matter!
    } elsif (/^(--target|-T)$/) {
        $call_target = shift @ARGV;
    } elsif (/^(--target=|-T)(.+)$/) {
        $call_target = $2;
    } elsif (/^--as-root$/) {
        $call_target_as_root = 1;
    } elsif (/^-nc$/) {
	$noclean = 1;
    } elsif (/^-b$/) {
	$binaryonly = '-b';
	@checkbuilddep_args = ();
	$binarytarget = 'binary';
    } elsif (/^-B$/) {
	$binaryonly = '-B';
	@checkbuilddep_args = ('-B');
	$binarytarget = 'binary-arch';
    } elsif (/^-A$/) {
	$binaryonly = '-A';
	@checkbuilddep_args = ();
	$binarytarget = 'binary-indep';
    } elsif (/^-S$/) {
	$sourceonly = '-S';
	@checkbuilddep_args = ('-B');
    } elsif (/^-v(.*)$/) {
	$since = $1;
    } elsif (/^-m(.*)$/) {
	$maint = $1;
    } elsif (/^-e(.*)$/) {
	$changedby = $1;
    } elsif (/^-C(.*)$/) {
	$desc = $1;
    } elsif (m/^-[EW]$/) {
	# Deprecated option
	warning(_g("-E and -W are deprecated, they are without effect"));
    } elsif (/^-R(.*)$/) {
	@debian_rules = split /\s+/, $1;
    } else {
	usageerr(_g("unknown option or argument %s"), $_);
    }
}

if ($binaryonly and $sourceonly) {
    usageerr(_g("cannot combine %s and %s"), $binaryonly, $sourceonly);
}
if ($noclean) {
    # -nc without -b/-B/-A/-S implies -b
    $binaryonly = '-b' unless ($binaryonly or $sourceonly);
}

if ($< == 0) {
    warning(_g("using a gain-root-command while being root")) if (@rootcommand);
} else {
    push @rootcommand, "fakeroot" unless @rootcommand;

    if (!testcommand($rootcommand[0])) {
	if ($rootcommand[0] eq 'fakeroot') {
	    error(_g("fakeroot not found, either install the fakeroot\n" .
	             "package, specify a command with the -r option, " .
	             "or run this as root"));
	} else {
	    error(_g("gain-root-commmand '%s' not found"), $rootcommand[0]);
	}
    }
}

unless ($signcommand) {
    $signsource = 0;
    $signchanges = 0;
}

my $signinterface;
if ($forcesigninterface) {
    $signinterface = $forcesigninterface;
} else {
    $signinterface = $signcommand;
}

if ($signcommand) {
    if ($signinterface !~ /^(gpg|pgp)$/) {
	warning(_g("unknown sign command, assuming pgp style interface"));
    } elsif ($signinterface eq 'pgp') {
	if ($signsource or $signchanges) {
	    warning(_g("PGP support is deprecated (see README.feature-removal-schedule)"));
	}
    }
}

my $build_opts = Dpkg::BuildOptions::parse();
if (defined $parallel) {
    $parallel = $build_opts->{parallel} if exists $build_opts->{parallel};
    $ENV{MAKEFLAGS} ||= '';
    $ENV{MAKEFLAGS} .= " -j$parallel";
    $build_opts->{parallel} = $parallel;
    Dpkg::BuildOptions::set($build_opts);
}

my $default_flags = exists $build_opts->{noopt} ? "-g -O0" : "-g -O2";
my %flags = ( CPPFLAGS => '',
	      CFLAGS   => $default_flags,
	      CXXFLAGS => $default_flags,
	      FFLAGS   => $default_flags,
	      LDFLAGS  => '-Wl,-Bsymbolic-functions',
    );

foreach my $flag (keys %flags) {
    if ($ENV{$flag}) {
	printf(_g("%s: use %s from environment: %s\n"), $progname,
		  $flag, $ENV{$flag});
    } else {
	$ENV{$flag} = $flags{$flag};
	printf(_g("%s: set %s to default value: %s\n"), $progname,
		  $flag, $ENV{$flag});
    }
    if ($ENV{"${flag}_APPEND"}) {
	$ENV{$flag} .= " ".$ENV{"${flag}_APPEND"};
    }
}

# Allow control of hardening-wrapper via dpkg-buildpackage DEB_BUILD_OPTIONS
my $hardening = $build_opts->{'hardening'};
if (defined $build_opts->{'nohardening'}) {
    $hardening = 0;
}
if (defined $hardening) {
    my $flag = 'DEB_BUILD_HARDENING';
    if ($hardening ne "0") {
        if (! -x '/usr/bin/hardened-cc') {
            syserr(_g("%s: 'hardening' flag found but 'hardening-wrapper' not installed\n"),
                $progname);
        }
        if ($hardening ne "1") {
            my @options = split(/,\s*/,$hardening);
            $hardening=1;

            my @hardopts = ('format', 'fortify', 'stackprotector',
                            'pie', 'relro');
            foreach my $item (@hardopts) {
                my $upitem = uc($item);
                foreach my $option (@options) {
                    if ($option =~ /^(no)?$item$/) {
                        $ENV{$flag.'_'.$upitem} = ($1 eq "");
                    }
                }
            }
        }
    }
    if (defined $ENV{$flag}) {
        printf(_g("%s: overriding %s in environment: %s\n"), $progname,
                $flag, $hardening);
    }
    $ENV{$flag}=$hardening;
}

my $cwd = cwd();
my $dir = basename($cwd);

my $changelog = changelog_parse();

my $pkg = mustsetvar($changelog->{source}, _g('source package'));
my $version = mustsetvar($changelog->{version}, _g('source version'));
my ($ok, $error) = version_check($version);
error($error) unless $ok;

my $maintainer;
if ($changedby) {
    $maintainer = $changedby;
} elsif ($maint) {
    $maintainer = $maint;
} else {
    $maintainer = mustsetvar($changelog->{maintainer}, _g('source changed by'));
}

open my $arch_env, '-|', 'dpkg-architecture', "-a$targetarch",
    "-t$targetgnusystem", '-s', '-f' or subprocerr('dpkg-architecture');
while ($_ = <$arch_env>) {
    chomp;
    my @cmds = split /\s*;\s*/;
    foreach (@cmds) {
	/^\s*(\w+)=([\w-]*)\s*$/ && do {
	    $ENV{$1} = $2;
	};
    }
}
close $arch_env or subprocerr('dpkg-architecture');

# In case of cross-compilation, give sensible default search path
# for some widely used tools
$targetgnusystem = debarch_to_gnutriplet($targetarch) if $targetarch;
if ($targetgnusystem and
    ($targetgnusystem ne debarch_to_gnutriplet(get_build_arch())))
{
   $ENV{PKG_CONFIG_LIBDIR} ||= "/usr/$targetgnusystem/lib/pkgconfig/:" .
                               "/usr/share/pkgconfig/";
}

my $arch;
unless ($sourceonly) {
    $arch = mustsetvar($ENV{'DEB_HOST_ARCH'}, _g('host architecture'));
} else {
    $arch = 'source';
}

# Preparation of environment stops here

(my $sversion = $version) =~ s/^\d+://;

my $pv = "${pkg}_$sversion";
my $pva = "${pkg}_${sversion}_$arch";

if (not -x "debian/rules") {
    warning(_g("debian/rules is not executable: fixing that."));
    chmod(0755, "debian/rules"); # No checks of failures, non fatal
}

if ($checkbuilddep) {
    if ($admindir) {
	push @checkbuilddep_args, "--admindir=$admindir";
    }

    system('dpkg-checkbuilddeps', @checkbuilddep_args);
    if (not WIFEXITED($?)) {
        subprocerr('dpkg-checkbuilddeps');
    } elsif (WEXITSTATUS($?)) {
	warning(_g("Build dependencies/conflicts unsatisfied; aborting."));
	warning(_g("(Use -d flag to override.)"));

	if ($sourceonly) {
	    warning(_g("This is currently a non-fatal warning with -S, but"));
	    warning(_g("will probably become fatal in the future."));
	} else {
	    exit 3;
	}
    }
}

if ($call_target) {
    if ($call_target_as_root or
        $call_target =~ /^(clean|binary(|-arch|-indep))$/)
    {
        withecho(@rootcommand, @debian_rules, $call_target);
    } else {
        withecho(@debian_rules, $call_target);
    }
    exit 0;
}

unless ($noclean) {
    withecho(@rootcommand, @debian_rules, 'clean');
}
unless ($binaryonly) {
    warning(_g("it is a bad idea to generate a source package " .
               "without cleaning up first, it might contain undesired " .
               "files.")) if $noclean;
    chdir('..') or syserr('chdir ..');
    my @opts = @passopts;
    if ($diffignore) { push @opts, $diffignore }
    push @opts, @tarignore;
    withecho('dpkg-source', @opts, '-b', $dir);
    chdir($dir) or syserr("chdir $dir");
}
unless ($sourceonly) {
    withecho(@debian_rules, 'build');
    withecho(@rootcommand, @debian_rules, $binarytarget);
}
if ($usepause &&
    ($signchanges || ( !$binaryonly && $signsource )) ) {
    print _g("Press the return key to start signing process\n");
    getc();
}

my $signerrors;
unless ($binaryonly) {
    if ($signsource && signfile("$pv.dsc")) {
	$signerrors = _g("Failed to sign .dsc and .changes file");
	$signchanges = 0;
    }
}

my @change_opts;

if ($binaryonly) { push @change_opts, $binaryonly }
if ($sourceonly) { push @change_opts, $sourceonly }
if ($sourcestyle) { push @change_opts, $sourcestyle }

if (defined($maint)) { push @change_opts, "-m$maint" }
if (defined($changedby)) { push @change_opts, "-e$changedby" }
if (defined($since)) { push @change_opts, "-v$since" }
if (defined($desc)) { push @change_opts, "-C$desc" }

my $chg = "../$pva.changes";
print STDERR " dpkg-genchanges @change_opts >$chg\n";
open CHANGES, '-|', 'dpkg-genchanges', @change_opts
    or subprocerr('dpkg-genchanges');

open OUT, '>', $chg or syserr(_g('write changes file'));

my $infiles = my $files = '';
while ($_ = <CHANGES>) {
    print OUT $_ or syserr(_g('write changes file'));
    chomp;

    if (/^Files:/i) {
	$infiles = 1;
    } elsif ($infiles && /^\s+(.*)$/) {
	$files .= " $1 ";
    } elsif ($infiles && /^\S/) {
	$infiles = 0;
    }
}

close CHANGES or subprocerr(_g('dpkg-genchanges'));
close OUT or syserr(_g('write changes file'));

my $srcmsg;
sub fileomitted($) { return $files !~ /$_[0]/ }
if (fileomitted '\.deb') {
    # source only upload
    if (fileomitted "\.diff\.$comp_regex" and fileomitted "\.debian\.tar\.$comp_regex") {
	$srcmsg = _g('source only upload: Debian-native package');
    } elsif (fileomitted "\.orig\.tar\.$comp_regex") {
	$srcmsg = _g('source only, diff-only upload (original source NOT included)');
    } else {
	$srcmsg = _g('source only upload (original source is included)');
    }
} else {
    $srcmsg = _g('full upload (original source is included)');
    if (fileomitted '\.dsc') {
	$srcmsg = _g('binary only upload (no source included)');
    } elsif (fileomitted "\.diff\.$comp_regex" and fileomitted "\.debian\.tar\.$comp_regex") {
	$srcmsg = _g('full upload; Debian-native package (full source is included)');
    } elsif (fileomitted "\.orig\.tar\.$comp_regex") {
	$srcmsg = _g('binary and diff upload (original source NOT included)');
    } else {
	$srcmsg = _g('full upload (original source is included)');
    }
}

if ($signchanges && signfile("$pva.changes")) {
    $signerrors = _g("Failed to sign .changes file");
}

if ($cleansource) {
    withecho(@rootcommand, @debian_rules, 'clean');
}

print "$progname: $srcmsg\n";
if ($signerrors) {
    warning($signerrors);
    exit 1;
}

sub testcommand {
    my ($cmd) = @_;

    my $fullcmd = `which $cmd`;
    chomp $fullcmd;
    return $fullcmd && -x $fullcmd;
}

sub mustsetvar {
    my ($var, $text) = @_;

    error(_g("unable to determine %s"), $text)
	unless defined($var);

    print "$progname: $text $var\n";
    return $var;
}

sub withecho {
    shift while !$_[0];
    print STDERR " @_\n";
    system(@_)
	and subprocerr("@_");
}

sub signfile {
    my ($file) = @_;
    print STDERR " signfile $file\n";
    my $qfile = quotemeta($file);

    if ($signinterface eq 'gpg') {
	system("(cat ../$qfile ; echo '') | ".
	       "$signcommand --utf8-strings --local-user "
	       .quotemeta($signkey||$maintainer).
	       " --clearsign --armor --textmode  > ../$qfile.asc");
    } else {
	system("$signcommand -u ".quotemeta($signkey||$maintainer).
	       " +clearsig=on -fast <../$qfile >../$qfile.asc");
    }
    my $status = $?;
    unless ($status) {
	system('mv', '--', "../$file.asc", "../$file")
	    and subprocerr('mv');
    } else {
	system('rm', '-f', "../$file.asc")
	    and subprocerr('rm -f');
    }
    print "\n";
    return $status
}
